### PR TITLE
Allow users via SMS

### PR DESCRIPTION
**Pull request by isdavid**
#### Also includes:
- Ability to get contact number
- Marshmallow permission requests for SMS and contacts
- Simple notification when getting SMS and being allowed
#### Potential issue for marshmallow users:

SMS permissions are not granted until trying to send SMS. This means user might not be able to receive a SMS until this has been done. Not a huge issue, but something to keep in mind. Will need to test with @gopenshaw. I don't think this issue should stop this PR.

### CODE REVIEW

**Comment by isdavid** [comment]
Thanks for pointing out the notification not clearing. I'll add that real quick.

File: app/src/main/java/org/island/messaging/MessageLayer.java [hunk 1]
```diff
@@ -137,7 +137,7 @@ private static void addFriendToDatabaseAndCreateDefaultProfile(Context context,
         }
     }
 
-    public static String getQrCode(Context context) {
+    public static String getEncodedString(Context context) {
```
Comment (gopenshaw) [inline app/src/main/java/org/island/messaging/MessageLayer.java]: I think `getEncodedString` might be too general. Maybe something like `getMyEncodedIdentity` or `getEncodedIdentityString`?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -61,6 +61,14 @@
     <string name="qr_action_dialog">Pick QR Action</string>
     <string name="title_activity_show_qr">ShowQRActivity</string>
     <string name="pseudonym_key_id">pseudonym_key_id</string>
+    <string name="sms_prefix">ISLAND-#-#-</string>
+    <string name="sms_allow_dialog_title">Allow user with SMS</string>
+    <string name="sms_disclaimer">Disclaimer: Standard text messaging rates may apply</string>
```
Comment (gopenshaw) [inline app/src/main/res/values/strings.xml:66]: Nice disclaimer :+1:

File: app/src/main/java/com/island/island/Activities/FeedActivity.java [hunk 3]
```diff
@@ -46,6 +58,15 @@
     private RecyclerView.LayoutManager mLayoutManager;
     private List<Post> mArrayOfPosts;
     private SwipeRefreshLayout refreshLayout;
+    private CoordinatorLayout mainLayout;
```
Comment (gopenshaw) [inline app/src/main/java/com/island/island/Activities/FeedActivity.java:61]: I noticed the lower case m prefix for member variables on some of our code and the stuff in the Udacity course. In this class we are using it only sometimes. I am not familiar with Android conventions to know if it's for all variables or only certain types of variables.

File: app/src/main/java/com/island/island/SmsReceiver.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            newFriendNotification(context);
+        }
+    }
+
+    private boolean isRelevant(Context context, String smsBody)
+    {
+        return smsBody.startsWith(context.getString(R.string.sms_prefix));
+    }
+
+    // TODO: Pass in user name
```
Comment (gopenshaw) [inline app/src/main/java/com/island/island/SmsReceiver.java:75]: Can we get this TODO now or should it wait until later? I'm OK either way :palm_tree:

**Comment by gopenshaw** [comment]
Overall this is super awesome. I think this will be a really useful feature! None of the issues I commented on are super critical. I would say merge it in whenever you feel good about it :sunglasses:

File: app/src/main/java/com/island/island/Activities/FeedActivity.java (hunk 3 above)
Comment (isdavid) [inline app/src/main/java/com/island/island/Activities/FeedActivity.java:61]: I don't think there's a preferred convention, I'm just inconsistent :)

File: app/src/main/java/com/island/island/SmsReceiver.java (hunk 4 above)
Comment (isdavid) [inline app/src/main/java/com/island/island/SmsReceiver.java:75]: I resolved this by returning a Boolean from addFriendFromEncodedString. True if a friend was actually added.

**Comment by isdavid** [comment]
I think I'll leave the naming conventions be for this PR. I'm planning sometime soon to do a code clean up PR. There I will fix naming and style inconsistencies.

**Pull request merged**